File: tailor_shop/src/main/java/com/example/tailor_shop/modules/flashsale/service/FlashSaleService.java
package com.example.tailor_shop.modules.flashsale.service;

import com.example.tailor_shop.modules.flashsale.dto.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Flash Sale Service Interface
 */
public interface FlashSaleService {

    // ==================== PUBLIC QUERIES ====================

    /**
     * Get currently active flash sales
     */
    List<FlashSaleResponse> getActiveSales(Long userId);

    /**
     * Get upcoming flash sales
     */
    List<FlashSaleResponse> getUpcomingSales(Long userId);

    /**
     * Get featured flash sales
     */
    List<FlashSaleResponse> getFeaturedSales(Long userId);

    /**
     * Get flash sale detail
     */
    FlashSaleResponse getDetail(Long id, Long userId);

    // ==================== ADMIN OPERATIONS ====================

    /**
     * List all flash sales (admin)
     */
    Page<FlashSaleResponse> listAll(List<String> statuses, Pageable pageable);

    /**
     * Create flash sale (admin)
     */
    FlashSaleResponse create(FlashSaleRequest request, Long createdBy);

    /**
     * Update flash sale (admin)
     */
    FlashSaleResponse update(Long id, FlashSaleRequest request, Long updatedBy);

    /**
     * Cancel flash sale (admin)
     */
    FlashSaleResponse cancel(Long id, Long updatedBy);

    // ==================== CUSTOMER OPERATIONS ====================

    /**
     * Purchase flash sale item - THE CORE ALGORITHM
     */
    FlashSalePurchaseResult purchase(Long flashSaleId, FlashSalePurchaseRequest request, Long userId);

    /**
     * Get my orders for a flash sale
     */
    List<FlashSaleOrderResponse> getMyOrders(Long flashSaleId, Long userId);

    /**
     * Get all my flash sale orders
     */
    Page<FlashSaleOrderResponse> getAllMyOrders(Long userId, Pageable pageable);

    /**
     * Get order detail
     */
    FlashSaleOrderResponse getOrderDetail(Long orderId, Long userId);

    /**
     * Cancel order (before payment)
     */
    FlashSaleOrderResponse cancelOrder(Long orderId, Long userId);

    /**
     * Confirm payment (simple flow without payment gateway)
     */
    FlashSaleOrderResponse confirmPayment(Long orderId, Long userId, String paymentMethod);

    // ==================== SCHEDULED TASKS ====================

    /**
     * Update flash sale statuses (SCHEDULED -> ACTIVE -> ENDED)
     */
    void updateFlashSaleStatuses();

    /**
     * Release expired reservations
     */
    void releaseExpiredReservations();

    /**
     * Expire pending orders past deadline
     */
    void expirePendingOrders();
}

File: frontend/src/services/flashSaleService.js
import api from './api/httpClient';

/**
 * Flash Sale Service - API Client
 */
const flashSaleService = {
    // ==================== PUBLIC ENDPOINTS ====================

    /**
     * Get active flash sales
     */
    getActiveSales: async () => {
        const response = await api.get('/flash-sales/active');
        return response.data;
    },

    /**
     * Get upcoming flash sales
     */
    getUpcomingSales: async () => {
        const response = await api.get('/flash-sales/upcoming');
        return response.data;
    },

    /**
     * Get featured flash sales
     */
    getFeaturedSales: async () => {
        const response = await api.get('/flash-sales/featured');
        return response.data;
    },

    /**
     * Get flash sale detail
     */
    getDetail: async (id) => {
        const response = await api.get(`/flash-sales/${id}`);
        return response.data;
    },

    // ==================== ADMIN ENDPOINTS ====================

    /**
     * List all flash sales (admin)
     */
    listAll: async (params = {}) => {
        const response = await api.get('/flash-sales', { params });
        return response.data;
    },

    /**
     * Create flash sale (admin)
     */
    create: async (data) => {
        const response = await api.post('/flash-sales', data);
        return response.data;
    },

    /**
     * Update flash sale (admin)
     */
    update: async (id, data) => {
        const response = await api.put(`/flash-sales/${id}`, data);
        return response.data;
    },

    /**
     * Cancel flash sale (admin)
     */
    cancel: async (id) => {
        const response = await api.post(`/flash-sales/${id}/cancel`);
        return response.data;
    },

    // ==================== CUSTOMER ENDPOINTS ====================

    /**
     * Purchase flash sale item
     */
    purchase: async (flashSaleId, data) => {
        const response = await api.post(`/flash-sales/${flashSaleId}/purchase`, data);
        return response.data;
    },

    /**
     * Get my orders for a flash sale
     */
    getMyOrdersForSale: async (flashSaleId) => {
        const response = await api.get(`/flash-sales/${flashSaleId}/my-orders`);
        return response.data;
    },

    /**
     * Get all my flash sale orders
     */
    getAllMyOrders: async (params = {}) => {
        const response = await api.get('/flash-sales/my-orders', { params });
        return response.data;
    },

    /**
     * Get order detail
     */
    getOrderDetail: async (orderId) => {
        const response = await api.get(`/flash-sales/orders/${orderId}`);
        return response.data;
    },

    /**
     * Cancel order
     */
    cancelOrder: async (orderId) => {
        const response = await api.post(`/flash-sales/orders/${orderId}/cancel`);
        return response.data;
    },

    /**
     * Confirm payment
     */
    confirmPayment: async (orderId, paymentMethod) => {
        const response = await api.post(`/flash-sales/orders/${orderId}/confirm-payment`, null, {
            params: { paymentMethod }
        });
        return response.data;
    }
};

export default flashSaleService;
